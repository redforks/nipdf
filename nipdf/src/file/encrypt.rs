use crate::object::ObjectId;
use ahash::HashMap;
use arc4::Arc4;
use log::error;
use md5::{Digest, Md5};
use nipdf_macro::{pdf_object, TryFromIntObject, TryFromNameObject};
use prescript::{sname, Name};
use tinyvec::{Array, ArrayVec, TinyVec};

#[derive(TryFromIntObject, Default, Debug, PartialEq, Eq, Clone, Copy)]
pub enum Algorithm {
    #[default]
    Undocument = 0,
    Key40 = 1,
    Key40AndMore = 2,
    Unpublished = 3,
    DefinedInDoc = 4,
}

#[derive(TryFromIntObject, PartialEq, Eq, Debug, Clone, Copy, PartialOrd, Ord)]
pub enum StandardHandlerRevion {
    V2 = 2,
    V3 = 3,
    V4 = 4,
}

#[derive(TryFromNameObject, PartialEq, Eq, Debug, Clone, Copy, PartialOrd, Ord, Default)]
pub enum DecryptMethod {
    #[default]
    None,
    V2,
    AesV2,
}

const fn identity() -> Name {
    sname("Identity")
}

#[pdf_object(Some("CryptFilter"))]
pub trait CryptFilterDictTrait {
    #[key("CFM")]
    #[try_from]
    fn decrypt_method(&self) -> DecryptMethod;
}

#[pdf_object(())]
pub trait EncryptDictTrait {
    fn filter(&self) -> Name;

    fn sub_filter(&self) -> Option<Name>;

    #[or_default]
    #[key("V")]
    #[try_from]
    fn algorithm(&self) -> Algorithm;

    #[key("Length")]
    #[default(40)]
    fn key_length(&self) -> u32;

    #[key("P")]
    fn permission_flags(&self) -> u32;

    #[key("R")]
    #[try_from]
    fn revison(&self) -> StandardHandlerRevion;

    /// 32-byte long string.
    #[key("O")]
    fn owner_password_hash(&self) -> &[u8];

    /// 32-byte long string.
    #[key("U")]
    fn user_password_hash(&self) -> &[u8];

    #[key("CF")]
    #[one_or_more]
    #[nested]
    fn crypt_filter_params(&self) -> HashMap<Name, CryptFilterDict>;

    /// Crypt filter used if Crypt field not set on Stream dictionary
    #[key("StmF")]
    #[default_fn(identity)]
    fn stream_default_crypt_filter(&self) -> Name;

    /// Crypt filter used to decode strings
    #[key("StrF")]
    #[default_fn(identity)]
    fn string_crypt_filter(&self) -> Name;
}

impl<'a, 'b> EncryptDict<'a, 'b> {
    /// Return the default crypt filter for stream and string.
    /// If `self.>revision()` not V4, return (Identity, Identity).
    /// Stream default crypt filter: lookup `self.crypt_filter_params` by
    /// `self.stream_default_crypt_filter()`, If not found, use Identity, otherwise by
    /// `decrypt_method()` return value.
    /// String default crypt filter similar to stream, but lookup by
    /// `self.string_default_crypt_filter()`. If any error occurs, uses Identity.
    pub fn default_filters(&self) -> (CryptFilter, CryptFilter) {
        use anyhow::{anyhow, Result};
        fn resolve_filter(
            name: &Name,
            params: &HashMap<Name, CryptFilterDict>,
        ) -> Result<CryptFilter> {
            if name == &identity() {
                return Ok(CryptFilter::Identity);
            }

            params
                .get(name)
                .ok_or_else(|| anyhow!("crypt filter params not found: {}", name))
                .and_then(|d| d.decrypt_method())
                .map(|m| match m {
                    DecryptMethod::None => CryptFilter::Identity,
                    DecryptMethod::V2 => CryptFilter::Rc4,
                    DecryptMethod::AesV2 => CryptFilter::Aes,
                })
        }

        fn _do(this: &EncryptDict) -> Result<(CryptFilter, CryptFilter)> {
            if this.revison()? != StandardHandlerRevion::V4 {
                return Ok((CryptFilter::Identity, CryptFilter::Identity));
            }

            let params = this.crypt_filter_params()?;
            let stream_filter = resolve_filter(&this.stream_default_crypt_filter()?, &params)?;
            let string_filter = resolve_filter(&this.string_crypt_filter()?, &params)?;
            Ok((stream_filter, string_filter))
        }

        if !matches!(
            self.algorithm().unwrap(),
            Algorithm::Key40 | Algorithm::Key40AndMore,
        ) {
            todo!("Algorithm: {:?}", self.algorithm().unwrap());
        }

        _do(self).unwrap_or_else(|e| {
            error!("failed to get default crypt filter, use Identity: {}", e);
            (CryptFilter::Identity, CryptFilter::Identity)
        })
    }
}

const PADDING: [u8; 32] = [
    0x28u8, 0xBF, 0x4E, 0x5E, 0x4E, 0x75, 0x8A, 0x41, 0x64, 0x00, 0x4E, 0x56, 0xFF, 0xFA, 0x01,
    0x08, 0x2E, 0x2E, 0x00, 0xB6, 0xD0, 0x68, 0x3E, 0x80, 0x2F, 0x0C, 0xA9, 0xFE, 0x64, 0x53, 0x69,
    0x7A,
];

/// Pad or truncate the password to 32 bytes.
/// If the password is longer than 32 bytes, the extra bytes are ignored.
/// If the password is shorter than 32 bytes, it is padded with bytes
/// [28 BF 4E 5E 4E 75 8A 41 64 00 4E 56 FF FA 01 08 2E 2E 00 B6 D0 68 3E 80 2F 0C A9 FE 64 53 69
/// 7A]
fn pad_trunc_password(s: &[u8]) -> [u8; 32] {
    let mut iter = s.iter().copied().chain(PADDING).take(32);
    std::array::from_fn(|_| iter.next().unwrap())
}

// algorithm 2
pub fn calc_encrypt_key(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> Box<[u8]> {
    let user_password = pad_trunc_password(user_password);
    let mut md5 = Md5::new();
    md5.update(user_password);
    md5.update(owner_hash);
    md5.update(permission_flags.to_le_bytes());
    md5.update(doc_id);
    // md5.update(&[0xff, 0xff, 0xff, 0xff]);
    let mut hash = md5.finalize();
    let n = key_length / 8;
    if revion > StandardHandlerRevion::V2 {
        for _ in 0..50 {
            hash = Md5::digest(&hash[..n]);
        }
    }
    (&hash[..n]).into()
}

// algorithm 4 and 5
fn calc_user_hash(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> [u8; 32] {
    let key = calc_encrypt_key(
        revion,
        key_length,
        user_password,
        owner_hash,
        permission_flags,
        doc_id,
    );

    if revion == StandardHandlerRevion::V2 {
        let mut r = PADDING.to_owned();
        Arc4::with_key(&key).encrypt(&mut r);
        r
    } else {
        let mut md5 = Md5::new();
        md5.update(PADDING);
        md5.update(doc_id);
        let mut hash = md5.finalize();
        let mut tmp = key.to_vec();
        for i in 0..=19 {
            for (t, k) in tmp.as_mut_slice().iter_mut().zip(&key[..]) {
                *t = *k ^ i;
            }
            Arc4::with_key(&tmp[..]).encrypt(&mut hash);
        }
        let mut r = [0u8; 32];
        r[..16].copy_from_slice(&hash[..]);
        r
    }
}

// algorithm 6
pub fn authorize_user(
    revion: StandardHandlerRevion,
    key_length: usize,
    user_password: &[u8],
    owner_hash: &[u8; 32],
    user_hash: &[u8; 32],
    permission_flags: u32,
    doc_id: &[u8],
) -> bool {
    let hash = calc_user_hash(
        revion,
        key_length,
        user_password,
        owner_hash,
        permission_flags,
        doc_id,
    );

    if revion == StandardHandlerRevion::V2 {
        &hash[..] == user_hash
    } else {
        hash[..16] == user_hash[..16]
    }
}

/// algorithm 3 step a to d
#[allow(dead_code)]
fn calc_rc4_key(
    revion: StandardHandlerRevion,
    key_length: usize,
    owner_password: &[u8],
) -> Box<[u8]> {
    let mut owner_password = Md5::digest(pad_trunc_password(owner_password));
    if revion > StandardHandlerRevion::V2 {
        for _ in 0..50 {
            owner_password = Md5::digest(&owner_password[..]);
        }
    }
    (&owner_password[..(key_length / 8)]).into()
}

// algorithm 7
#[allow(dead_code)]
fn authorize_owner(
    revion: StandardHandlerRevion,
    key_length: usize,
    owner_password: &[u8],
    owner_hash: &[u8; 32],
    user_hash: &[u8; 32],
    permission_flag: u32,
    doc_id: &[u8],
) -> bool {
    let rc4_key = &calc_rc4_key(revion, key_length, owner_password);
    let mut decrypt = owner_hash.to_vec();
    if revion == StandardHandlerRevion::V2 {
        Arc4::with_key(rc4_key).encrypt(&mut decrypt);
    } else {
        let mut tmp = rc4_key.to_vec();
        for i in (0..19).rev() {
            for (t, k) in tmp.as_mut_slice().iter_mut().zip(&rc4_key[..]) {
                *t = *k ^ i;
            }
            Arc4::with_key(&tmp[..]).encrypt(&mut decrypt);
        }
    }

    authorize_user(
        revion,
        key_length,
        &decrypt,
        owner_hash,
        user_hash,
        permission_flag,
        doc_id,
    )
}

pub trait VecLike {
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]>;
    fn as_mut_slice(&mut self) -> &mut [u8];
}

pub trait Decryptor {
    fn new(key: &[u8], id: ObjectId) -> Self;
    fn decrypt<V: VecLike>(&self, data: &mut V);
}

impl<A> VecLike for TinyVec<A>
where
    A: Array<Item = u8>,
{
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]> {
        self.drain(range).collect()
    }

    fn as_mut_slice(&mut self) -> &mut [u8] {
        self.as_mut_slice()
    }
}

impl VecLike for Vec<u8> {
    fn drain(&mut self, range: std::ops::Range<usize>) -> ArrayVec<[u8; 16]> {
        self.drain(range).collect()
    }

    fn as_mut_slice(&mut self) -> &mut [u8] {
        self.as_mut_slice()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CryptFilter {
    Identity,
    Rc4,
    Aes,
}

pub struct Rc4Decryptor(ArrayVec<[u8; 16]>);

impl Decryptor for Rc4Decryptor {
    fn new(key: &[u8], id: ObjectId) -> Self {
        let n = key.len();
        let mut k = TinyVec::<[u8; 16 + 5]>::with_capacity(n + 5);
        k.extend_from_slice(key);
        k.extend_from_slice(&id.id().0.to_le_bytes()[..3]);
        k.extend_from_slice(&id.generation().to_le_bytes()[..]);
        let key = Md5::digest(&k[..]);
        let key = key.into_iter().take((n + 5).min(16)).collect();
        Self(key)
    }

    fn decrypt<V: VecLike>(&self, data: &mut V) {
        Arc4::with_key(&self.0).encrypt(data.as_mut_slice());
    }
}

pub struct AesDecryptor(ArrayVec<[u8; 16]>);

impl Decryptor for AesDecryptor {
    fn new(key: &[u8], id: ObjectId) -> Self {
        let n = key.len();
        let mut k = TinyVec::<[u8; 16 + 5 + 4]>::with_capacity(n + 5 + 4);
        k.extend_from_slice(key);
        k.extend_from_slice(&id.id().0.to_le_bytes()[..3]);
        k.extend_from_slice(&id.generation().to_le_bytes()[..]);
        k.extend_from_slice(b"sAlT");
        let key = Md5::digest(&k[..]);
        let key = key.into_iter().take((n + 5).min(16)).collect();
        Self(key)
    }

    /// Decode data using aes, work in cbc mode,block size is 16 or Aes128.
    /// the initialization vector is a 16-byte random  number that is stored as the first 16 bytes
    /// of the encrypted data.
    /// Pad the data using the PKCS#5 padding scheme.
    fn decrypt<V: VecLike>(&self, data: &mut V) {
        use aes::cipher::{block_padding::Pkcs7, BlockDecryptMut, KeyIvInit};
        type Aes128CbcDec = cbc::Decryptor<aes::Aes128>;

        let iv_data = data.drain(0..16);
        let mut iv = [0u8; 16];
        iv.copy_from_slice(&iv_data[..]);
        Aes128CbcDec::new(self.0.as_ref().into(), &iv.into())
            .decrypt_padded_mut::<Pkcs7>(data.as_mut_slice())
            .unwrap();
    }
}

#[cfg(test)]
mod tests;
