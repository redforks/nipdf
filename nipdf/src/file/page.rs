use self::paint::Render;
pub use self::paint::{RenderOption, RenderOptionBuilder};
use crate::{
    graphics::{
        parse_operations, shading::ShadingDict, trans::FormToUserSpace, ColorArgs, ColorSpaceArgs,
        LineCapStyle, LineJoinStyle, Operation, PatternDict, Point, RenderingIntent,
    },
    object::{Dictionary, Object, ObjectValueError, PdfObject, Stream},
    text::FontDict,
};
use ahash::{HashMap, HashMapExt};
use educe::Educe;
use log::error;
use nipdf_macro::{pdf_object, TryFromNameObject};
use nom::Finish;
use prescript::Name;
use prescript_macro::name;
use std::iter::once;
use tiny_skia::Pixmap;

mod paint;

#[derive(Debug, Copy, Clone, PartialEq)]
pub struct Rectangle {
    pub left_x: f32,
    pub lower_y: f32,
    pub right_x: f32,
    pub upper_y: f32,
}

impl Rectangle {
    /// From left, top, right, bottom, re-order them to make sure that
    /// left <= right, top <= bottom
    pub fn from_lbrt(left_x: f32, bottom_y: f32, right_x: f32, top_y: f32) -> Self {
        Self {
            left_x: left_x.min(right_x),
            lower_y: bottom_y.min(top_y),
            right_x: left_x.max(right_x),
            upper_y: bottom_y.max(top_y),
        }
    }

    pub fn from_xywh(x: f32, y: f32, w: f32, h: f32) -> Self {
        Self::from_lbrt(x, y, x + w, y + h)
    }

    pub fn width(&self) -> f32 {
        self.right_x - self.left_x
    }

    pub fn height(&self) -> f32 {
        self.upper_y - self.lower_y
    }

    pub fn left_lower(&self) -> Point {
        Point::new(self.left_x, self.lower_y)
    }

    pub fn right_upper(&self) -> Point {
        Point::new(self.right_x, self.upper_y)
    }
}

impl From<Rectangle> for tiny_skia::Rect {
    fn from(rect: Rectangle) -> Self {
        Self::from_ltrb(rect.left_x, rect.lower_y, rect.right_x, rect.upper_y).unwrap()
    }
}

/// Convert from raw array, auto re-order to (left_x, lower_y, right_x, upper_y),
/// see PDF 32000-1:2008 7.9.5
impl TryFrom<&Object> for Rectangle {
    type Error = ObjectValueError;

    fn try_from(object: &Object) -> Result<Self, Self::Error> {
        match object {
            Object::Array(arr) => {
                let mut iter = arr.iter();
                let left_x = iter.next().unwrap().as_number().unwrap();
                let lower_y = iter.next().unwrap().as_number().unwrap();
                let right_x = iter.next().unwrap().as_number().unwrap();
                let upper_y = iter.next().unwrap().as_number().unwrap();
                Ok(Self::from_lbrt(left_x, lower_y, right_x, upper_y))
            }
            _ => Err(ObjectValueError::GraphicsOperationSchemaError),
        }
    }
}

#[pdf_object(Some("ExtGState"))]
pub trait GraphicsStateParameterDictTrait {
    #[key("LW")]
    fn line_width(&self) -> Option<f32>;

    #[key("LC")]
    #[try_from]
    fn line_cap(&self) -> Option<LineCapStyle>;

    #[key("LJ")]
    #[try_from]
    fn line_join(&self) -> Option<LineJoinStyle>;

    #[key("ML")]
    fn miter_limit(&self) -> Option<f32>;

    #[key("RI")]
    #[try_from]
    fn rendering_intent(&self) -> Option<RenderingIntent>;

    #[key("SA")]
    fn stroke_adjustment(&self) -> Option<bool>;

    #[key("CA")]
    fn stroke_alpha(&self) -> Option<f32>;
    #[key("ca")]
    fn fill_alpha(&self) -> Option<f32>;
    #[key("AIS")]
    fn alpha_source_flag(&self) -> Option<bool>;
    #[key("TK")]
    fn text_knockout_flag(&self) -> Option<bool>;
    #[key("FL")]
    fn flatness(&self) -> Option<f32>;
}

#[derive(Copy, Clone, PartialEq, Eq, Debug, TryFromNameObject)]
pub enum XObjectType {
    Image,
    Form,
    /// PostScript XObject
    PS,
}

#[pdf_object(Some("XObject"))]
pub trait XObjectDictTrait {
    #[try_from]
    fn subtype(&self) -> XObjectType;

    // available if it is soft-mask image, see Table 146
    #[try_from]
    fn matte(&self) -> Option<ColorArgs>;

    #[nested]
    fn s_mask(&self) -> Option<XObjectDict<'a, 'b>>;

    #[or_default]
    fn interpolate(&self) -> bool;

    #[or_default]
    fn image_mask(&self) -> bool;

    #[self_as]
    fn as_form(&self) -> FormXObjectDict<'a, 'b>;
}

impl<'a, 'b> XObjectDict<'a, 'b> {
    fn as_stream(&self) -> Result<&Stream, ObjectValueError> {
        self.d.resolver().resolve(self.id().unwrap())?.as_stream()
    }
}

#[pdf_object((Some("XObject"), "Form"))]
pub trait FormXObjectDictTrait {
    #[try_from]
    fn b_box(&self) -> Rectangle;

    #[try_from]
    #[or_default]
    fn matrix(&self) -> FormToUserSpace;

    #[nested]
    fn resources(&self) -> Option<ResourceDict<'a, 'b>>;

    fn group(&self) -> Option<&'b Dictionary>;

    #[key("Ref")]
    fn ref_dict(&self) -> Option<&'b Dictionary>;

    fn metadata(&self) -> Option<&'b Dictionary>;

    fn piece_info(&self) -> Option<&'b Dictionary>;
}

/// Wrap type to impl TryFrom<> trait
#[derive(Educe)]
#[educe(Deref)]
pub struct ColorSpaceResources(HashMap<Name, ColorSpaceArgs>);

impl<'b> TryFrom<&'b Object> for ColorSpaceResources {
    type Error = ObjectValueError;

    fn try_from(object: &'b Object) -> Result<Self, Self::Error> {
        let mut map = HashMap::new();
        match object {
            Object::Dictionary(dict) => {
                for (k, v) in dict.iter() {
                    let cs = ColorSpaceArgs::try_from(v)?;
                    map.insert(k.clone(), cs);
                }
                Ok(Self(map))
            }
            _ => {
                error!("{:?}", object);
                Err(ObjectValueError::GraphicsOperationSchemaError)
            }
        }
    }
}

#[pdf_object(())]
pub trait ResourceDictTrait {
    #[nested]
    fn ext_g_state(&self) -> HashMap<Name, GraphicsStateParameterDict<'a, 'b>>;
    #[try_from]
    fn color_space(&self) -> ColorSpaceResources;
    #[nested]
    fn pattern(&self) -> HashMap<Name, PatternDict<'a, 'b>>;
    #[nested]
    fn shading(&self) -> HashMap<Name, ShadingDict<'a, 'b>>;
    #[nested]
    fn x_object(&self) -> HashMap<Name, XObjectDict<'a, 'b>>;
    #[nested]
    fn font(&self) -> HashMap<Name, FontDict<'a, 'b>>;
    fn properties(&self) -> Option<&'b Dictionary>;
}

#[pdf_object(["Pages", "Page"])]
pub(crate) trait PageDictTrait {
    #[nested]
    fn kids(&self) -> Vec<Self>;
    #[try_from]
    fn media_box(&self) -> Option<Rectangle>;
    #[try_from]
    fn crop_box(&self) -> Option<Rectangle>;
    #[nested]
    fn resources(&self) -> Option<ResourceDict<'a, 'b>>;
    fn contents(&self) -> Vec<&Stream>;
    #[key("Type")]
    fn type_name(&self) -> &Name;
}

impl<'a, 'b> PageDict<'a, 'b> {
    pub fn is_leaf(&self) -> bool {
        self.type_name().unwrap() == &name!("Page")
    }
}

#[derive(Debug)]
pub struct Page<'a, 'b> {
    d: PageDict<'a, 'b>,
    parents_to_root: Vec<PageDict<'a, 'b>>,
}

impl<'a, 'b: 'a> Page<'a, 'b> {
    pub fn id(&self) -> u32 {
        self.d.id().unwrap().get()
    }

    fn iter_to_root(&self) -> impl Iterator<Item = &PageDict<'a, 'b>> {
        once(&self.d).chain(self.parents_to_root.iter())
    }

    pub fn media_box(&self) -> Rectangle {
        self.iter_to_root()
            .find_map(|d| d.media_box().unwrap())
            .expect("page must have media box")
    }

    pub fn crop_box(&self) -> Option<Rectangle> {
        self.iter_to_root().find_map(|d| d.crop_box().unwrap())
    }

    fn resources(&self) -> ResourceDict<'a, 'b> {
        self.iter_to_root()
            .find_map(|d| d.resources().unwrap())
            .expect("page must have resources")
    }

    pub fn content(&self) -> Result<PageContent, ObjectValueError> {
        let bufs = self
            .d
            .contents()
            .unwrap()
            .into_iter()
            .map(|s| s.decode(self.d.d.resolver()).map(|v| v.into_owned()))
            .collect::<Result<_, _>>()?;
        Ok(PageContent { bufs })
    }

    pub fn render_steps(
        &self,
        option: RenderOptionBuilder,
        steps: Option<usize>,
    ) -> Result<Pixmap, ObjectValueError> {
        let media_box = self.media_box();
        let crop_box = self.crop_box();
        let option = option
            .width(media_box.width() as u32)
            .height(media_box.height() as u32)
            .crop(need_crop(crop_box, media_box).then(|| crop_box.unwrap()))
            .build();
        let content = self.content()?;
        let ops = content.operations();
        let resource = self.resources();
        let mut canvas = option.create_canvas();
        let mut renderer = Render::new(&mut canvas, option, &resource);
        if let Some(steps) = steps {
            ops.into_iter().take(steps).for_each(|op| renderer.exec(op));
        } else {
            ops.into_iter().for_each(|op| renderer.exec(op));
        };
        drop(renderer);
        Ok(canvas)
    }

    pub fn render(&self, option: RenderOptionBuilder) -> Result<Pixmap, ObjectValueError> {
        self.render_steps(option, None)
    }

    /// Parse page tree to get all pages
    pub(crate) fn parse(root: PageDict<'a, 'b>) -> Result<Vec<Self>, ObjectValueError> {
        let mut pages = Vec::new();
        let mut parents = Vec::new();
        fn handle<'a, 'b: 'a, 'c>(
            node: PageDict<'a, 'b>,
            pages: &'c mut Vec<Page<'a, 'b>>,
            parents: &'c mut Vec<PageDict<'a, 'b>>,
        ) -> Result<(), ObjectValueError> {
            if node.is_leaf() {
                pages.push(Page::from_leaf(&node, &parents[..])?);
            } else {
                let kids = node.kids().unwrap();
                parents.push(node);
                for kid in kids {
                    handle(kid, pages, parents)?;
                }
            }
            Ok(())
        }
        handle(root, &mut pages, &mut parents)?;
        Ok(pages)
    }

    fn from_leaf(
        d: &PageDict<'a, 'b>,
        parents: &[PageDict<'a, 'b>],
    ) -> Result<Self, ObjectValueError> {
        let mut parents = parents.to_vec();
        parents.reverse();

        Ok(Self {
            d: d.clone(),
            parents_to_root: parents,
        })
    }
}

fn need_crop(crop: Option<Rectangle>, media: Rectangle) -> bool {
    match crop {
        None => false,
        Some(crop) => crop != media,
    }
}

pub struct PageContent {
    bufs: Vec<Vec<u8>>,
}

impl PageContent {
    pub fn new(bufs: Vec<Vec<u8>>) -> Self {
        Self { bufs }
    }

    pub fn operations(&self) -> Vec<Operation> {
        let mut r = vec![];
        for buf in &self.bufs {
            let (input, ops) = parse_operations(buf.as_ref()).finish().unwrap();
            assert!(input.is_empty(), "buf should be empty: {:?}", input);
            r.extend_from_slice(ops.as_slice());
        }
        r
    }

    pub fn as_ref(&self) -> impl Iterator<Item = &[u8]> {
        self.bufs.iter().map(|v| v.as_ref())
    }
}

#[cfg(test)]
mod tests;
