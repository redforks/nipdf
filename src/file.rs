//! Contains types of PDF file structures.
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Header<'a>(&'a [u8]);

impl<'a> Header<'a> {
    pub fn new(buf: &'a [u8]) -> Self {
        Self(buf)
    }

    pub fn ver(&self) -> &str {
        std::str::from_utf8(self.0).unwrap()
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct Tail {
    xref_pos: u32,
}

impl Tail {
    pub fn new(xref_pos: u32) -> Self {
        Self { xref_pos }
    }

    pub fn xref_pos(&self) -> u32 {
        self.xref_pos
    }
}
