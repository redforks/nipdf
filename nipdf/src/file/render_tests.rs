//! Test page render result using `insta` to ensure that the rendering result is not changed.
//! This file checks file pdfreference1.0.pdf
use crate::file::{File, RenderOptionBuilder};
use anyhow::Result as AnyResult;
use insta::assert_ron_snapshot;
use md5::{Digest, Md5};
use std::path::Path;

fn decode_file_page(path: &str, page_no: usize) -> AnyResult<String> {
    let path = Path::new(env!("CARGO_MANIFEST_DIR")).join(path);
    let buf = std::fs::read(&path)?;
    let f = File::parse(buf, "", "").unwrap_or_else(|e| panic!("failed to parse {path:?}: \n{e}"));
    let resolver = f.resolver()?;
    let catalog = f.catalog(&resolver)?;
    let pages = catalog.pages()?;
    let page = &pages[page_no];
    let option = RenderOptionBuilder::new().zoom(1.5);
    let bytes = page.render(option)?.into_vec();
    let hash = Md5::digest(&bytes[..]);
    Ok(hex::encode(hash))
}

/// Render page to image, and returns its md5 hash converted to hex
fn decode_page(page_no: usize) -> AnyResult<String> {
    decode_file_page("sample_files/normal/pdfreference1.0.pdf", page_no)
}

#[test]
fn clip_mask() {
    assert_ron_snapshot!(&decode_page(141).unwrap());
}

#[test]
fn mask_image() {
    assert_ron_snapshot!(&decode_page(163).unwrap());
}

#[test]
fn pattern_color() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/normal/SamplePdf1_12mb_6pages.pdf", 5).unwrap()
    );
}

#[test]
fn form() {
    assert_ron_snapshot!(&decode_file_page("sample_files/xobject/form.pdf", 0).unwrap());
}

#[test]
fn form_ctm() {
    assert_ron_snapshot!(
        &decode_file_page("../../pdf/ICEpower125ASX2_Datasheet_2.0.pdf", 4).unwrap()
    );
}

#[test]
fn radial_shade() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/bizarre/pdfReferenceUpdated.pdf", 809).unwrap()
    )
}

#[test]
fn axial_shade() {
    // TODO: find a sample page contains PaintShading("axial-shade") operation
    assert_ron_snapshot!(&decode_file_page("../../pdf/code.pdf", 619).unwrap())
}

#[test]
fn type0_cid_font() {
    assert_ron_snapshot!(
        &decode_file_page("sample_files/bizarre/pdfReferenceUpdated.pdf", 1013).unwrap()
    )
}

#[test]
fn standard_14_font_alias_name() {
    // Font name: TimesNewRomanPSMT alias of Times-Roman, see normalize_font_name()
    assert_ron_snapshot!(&decode_file_page("../../pdf/code.pdf", 620).unwrap())
}

#[test]
fn image_mask_cal_rgb_index_color_space() {
    // test paint image has mask, and its color space is Indexed to CalRGB,
    // image stream processed with Predicator
    assert_ron_snapshot!(&decode_file_page("sample_files/filters/predictor.pdf", 0).unwrap())
}

#[test]
fn decrypt_aes_revion3() {
    assert_ron_snapshot!(&decode_file_page("../../pdf/5176.CFF.pdf", 0).unwrap());
}

#[test]
fn ttf_font_cmap_trimmed_table_mapping() {
    // font used in graph, its cmap table uses format TrimmedTableMapping
    // that ttf-parser glyph_index() don't work, see `TTFParserFontOp::char_to_gid()`
    assert_ron_snapshot!(
        &decode_file_page("pdf.js/web/compressed.tracemonkey-pldi-09.pdf", 9).unwrap()
    )
}

#[test]
fn axial_shade_with_sample_function() {
    assert_ron_snapshot!(
        &decode_file_page("pdf.js/web/compressed.tracemonkey-pldi-09.pdf", 10).unwrap()
    )
}

#[test]
fn todo_rotate_n_encrypt_alg2() {
    // page rotate
    // encrypt algorithm 2 (Algorithm::Key40AndMore)
    // todo: ForceBold font flag should render glyph bolder
    assert_ron_snapshot!(&decode_file_page("../../pdf/avr-1507-owners-manual-en.pdf", 10).unwrap())
}

#[test]
fn todo_radius_pattern() {
    // todo: tiny_skia not support end radius, the page has both start and end radius rendered
    // incorrectly
    assert_ron_snapshot!(&decode_file_page("sample_files/bizarre/PDF32000_2008.pdf", 745).unwrap())
}

#[test]
fn todo_glyph_encoding_problem() {
    // todo: incorrect bullet glyph rendered, possible because ttf-parser glyph_index() returned
    // wrong glyph index
    assert_ron_snapshot!(&decode_file_page("sample_files/bizarre/PDF32000_2008.pdf", 159).unwrap())
}
