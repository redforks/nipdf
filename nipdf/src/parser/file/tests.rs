use super::*;
use crate::file::open_test_file;
use insta::assert_debug_snapshot;
use test_case::test_case;
use test_log::test;

#[test]
fn parse_file_header() {
    let buf = b"%PDF-1.7\n%comment\n";
    let (buf, header) = parse_header(buf).unwrap();
    assert_eq!(buf, b"");
    assert_eq!(header, "1.7");
}

#[test_case(None, b"hello", b"world"; "not exist")]
#[test_case(Some(0), b"hello", b"hello"; "matches")]
#[test_case(Some(1), b"\nhello", b"hello"; "after newline")]
#[test_case(Some(1), b"\nhello\n", b"hello"; "end with newline")]
#[test_case(Some(2), b"\r\nhello\r\n", b"hello"; "CRLF")]
#[test_case(Some(4), b"foo\nfoo\nbar", b"foo"; "from end")]
#[test_case(None, b"abc-foo", b"foo"; "not the whole line")]
fn test_r_find_start_object_tag(exp: Option<usize>, buf: &[u8], tag: &[u8]) {
    assert_eq!(exp, r_find_start_object_tag(buf, tag));
}

#[test]
fn test_parse_trailer() {
    let buf = b"trailer\n<< /Size 1 >>\nstartxref\n1234\n%%EOF";
    assert_debug_snapshot!(parse_trailer(buf).unwrap());
}

#[test_case(b"xref\n1 0\n", "empty")]
#[test_case(
    b"xref\n1 2\n0000000000 00000 n \n0000000010 00000 n \n",
    "two entries"
)]
fn test_parse_xref_table(buf: impl AsRef<[u8]>, name: &str) {
    assert_debug_snapshot!(name, parse_xref_table(buf.as_ref()).unwrap());
}

#[test]
fn test_parse_frame() {
    assert_debug_snapshot!(
        parse_frame(
            b"xref
1 2
0000000000 00000 n
0000000010 00000 n
trailer
<< /Size 1 >>
startxref
1234
%%EOF
"
        )
        .unwrap()
    );
}

#[test]
fn test_parse_frame_set() {
    assert_debug_snapshot!(
        parse_frame_set(
            b"%PDF-1.7
xref
1 1
0000000000 00000 n
trailer
<< /Size 1 >>
startxref
9
%%EOF
xref
1 1
0000000000 00000 n
trailer
<< /Prev 9 >>
startxref
77
%%EOF
"
        )
        .unwrap()
    );
}

#[test]
fn read_xref_stream() {
    let f = open_test_file("sample_files/file-structure/xref-stream.pdf");
    let resolver = f.resolver().unwrap();

    // assert object in file
    assert_eq!(
        sname("Catalog"),
        resolver
            .resolve(1)
            .unwrap()
            .as_dict()
            .unwrap()
            .get(&sname("Type"))
            .unwrap()
            .name()
            .unwrap()
    );

    // assert object in stream
    assert_eq!(
        1,
        resolver
            .resolve(2)
            .unwrap()
            .as_dict()
            .unwrap()
            .get(&sname("Count"))
            .unwrap()
            .int()
            .unwrap()
    );
}

#[test]
fn read_xref_stream_has_prev() {
    let f = open_test_file("pdf.js/test/pdfs/160F-2019.pdf");
    // get resolver without error
    f.resolver().unwrap();
}
