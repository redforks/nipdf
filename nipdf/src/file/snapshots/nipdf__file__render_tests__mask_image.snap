---
source: nipdf/src/file/render_tests.rs
expression: "&decode_page(163).unwrap()"
---
"f3d86a8b1a99eec01632a44f525d468b"
